cmds/tail/tail.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			return err
+		}
+		// merge this block to what read so far
+		readData = append(buf[:n], readData...)
+		// XXX this can be optimized, it's not necessary to re-count every time,
+		// just the last block
+		if bytes.Count(readData, []byte{'\n'}) >= numLines {
+			break
+		}
+	}
+	// find the start of the n-th to last line
+	slice := readData
+	// `data` contains up to `numLines` lines of the file
+	var data []byte
+	if len(slice) != 0 {
+		if slice[len(slice)-1] == '\n' {
+			// don't consider the last new line for the line count
+			slice = slice[:len(slice)-1]
+		}
+		var (
+			foundLines = 0
+			idx        int
+		)
+		for {
+			if foundLines >= numLines {
+				break
+			}
+			// find newlines backwards from the end of `slice`
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	_, err = writer.Write(data)
+	if err != nil {
+		return err
+	}
+	// reposition the stream at the end of what we've read
+	_, err = inFile.Seek(lastPos, os.SEEK_SET)
+	return nil
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile os.File, writer io.Writer, numLines int) error {
+	// get the file's block size. This will be the size of each chunk to read
+	blkSize, err := getBlockSize(inFile)
+	if err != nil {
+		return err
+	}
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
+	var slice []byte             // will hold the final data, after moving line by line
+	for {
+		n, err := io.ReadFull(&inFile, buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
+				return err
+			}
+		}
+		// look for newlines and keep a slice starting at the n-th to last line
+		// (no further than numLines)
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
+		foundNewLines := bytes.Count(slice, []byte{'\n'})
+		newLinesToSkip := 0
+		if foundNewLines > numLines {
+			newLinesToSkip = foundNewLines - numLines
+		}
+		// if there are more than `numLines` lines, skip them
+		for idx := 0; idx < newLinesToSkip; idx++ {
```

Comment:
I feel like this could be solved by using `IndexFunc`: https://play.golang.org/p/dJdHBHwsjbd

```go
m := bytes.IndexFunc(slice, NthByte(newLinesToSkip, '\n'))
if m == -1 {
   return fmt.Errorf("less bytes than expected were found")
}
slice = append(slice[m:], buf[:n])
```

Reply:
nice, I didn't know `IndexFunc`, thanks!

Reply:
I had a second thought - while I like the elegance of your solution, maybe it's not ideal this dance of converting to utf8 and making a function call for every rune (not byte). I'd rather stick with the original, yet less elegant code, unless you have a stronger reason not to

Reply:
why is it not ideal? utf8 will never give you a rune like "\xFF\xA" if that's what you're worried about. if there's a newline somewhere, it'll be a "\xA" rune.

Reply:
I was afraid of doing unnecessary conversions around, but as clarified with Ron this is not the case for Go. I will use `IndexFunc` :)